src/main/java/net/jqwik/micronaut/extension/JqwikMicronautExtension.java

```diff
@@ -152,6 +194,8 @@ private MicronautTestValue buildMicronautTestValue(final ContainerLifecycleConte
     }
 
     private MicronautTestValue buildValueObject(final JqwikMicronautTest micronautTest) {
+        PER_TRY = micronautTest.perTry();
+        JqwikMicronautRegistrar.registerHooks();
```

Comment:
I think this won't work in the general case.
1. You're using a global value to hold the `perTry` config which can differ on a per test class basis. Since registration of hooks and execution of tests is done at different times there is no guarantee that querying PER_TRY sees the correct value. If it does so for the current tests it's mere coincidence.
2. The same holds for the REGISTRAR global you are using. It just won't work.

What the jqwik extension/hooks model suggests to do is:
- Register all hooks that could be used, ie both the per-try and the per-property hooks
- Decide if you want to execute the body of the hook-method when it is invoked.

For example:
```java
public class InterceptBeforePropertyMethod ...
    public static class Pre implements AroundPropertyHook ...

    @Override
    @NonNullApi
    @Nonnull
    public PropertyExecutionResult aroundProperty(
            final PropertyLifecycleContext context,
            final PropertyExecutor property
    ) throws Exception {
        context.findAnnotation(JqwikMicronautTest.class).ifPresent(a -> {
            if (a.perTry()) {
                micronautExtension.preBeforePropertyMethod(context);
            }
        });
        return property.execute();
    }
```

I'm not sure if `findAnnotation` suffices or if you'll also have to use `findAnnotationsInContainer`.

Reply:
> I think this won't work in the general case.
> 
>     1. You're using a global value to hold the `perTry` config which can differ on a per test class basis. Since registration of hooks and execution of tests is done at different times there is no guarantee that querying PER_TRY sees the correct value. If it does so for the current tests it's mere coincidence.
> 
>     2. The same holds for the REGISTRAR global you are using. It just won't work.
> 
> 
> What the jqwik extension/hooks model suggests to do is:
> 
>     * Register all hooks that could be used, ie both the per-try and the per-property hooks
> 
>     * Decide if you want to execute the body of the hook-method when it is invoked.
> 
> 
> For example:
> 
> ```java
> public class InterceptBeforePropertyMethod ...
>     public static class Pre implements AroundPropertyHook ...
> 
>     @Override
>     @NonNullApi
>     @Nonnull
>     public PropertyExecutionResult aroundProperty(
>             final PropertyLifecycleContext context,
>             final PropertyExecutor property
>     ) throws Exception {
>         context.findAnnotation(JqwikMicronautTest.class).ifPresent(a -> {
>             if (a.perTry()) {
>                 micronautExtension.preBeforePropertyMethod(context);
>             }
>         });
>         return property.execute();
>     }
> ```
> 
> I'm not sure if `findAnnotation` suffices or if you'll also have to use `findAnnotationsInContainer`.

Thanks for the feedback. I updated the code accordingly and it's ready for you to review whenever you can.

If you accept any feedback on Jqwik's API, I really would love to have a common parent (that is, an `interface`) of both `PropertyLifecycleContext` and `TryLifecycleContext`.
That would greatly simplify extension's code by not having to duplicate methods depending on which one one wants to handle. You can see that in the form of:

``` java
if (context instanceof TryLifecycleContext) {
// ...
}

if (context instanceof PropertyLifecycleContext) {
// ...
}
```
I know that they are slightly different, but having one common ancestor would be awesome.

On the other hand, I noticed that the hooks for Properties and Tries are also slightly similar. I wish there is a way of doing something similar (probably the API is ok and it's me the one that can't think of a better approach to create hooks).

Anyway, I would love to get your feedback on the new changes. If that's ok, then I will think about on how to improve the code. One idea I had in mind was to introduce a lightweight DI framework such as Google Guice, but my priority now is to have the extension to do the right thing before improving it for readability/maintainability.

Reply:
> If you accept any feedback on Jqwik's API, I really would love to have a common parent (an `interface` would be great) of both `PropertyLifecycleContext` and `TryLifecycleContext`. That would greatly simplify extension's code by not having to duplicate methods depending on which one one wants to handle. You can see that in the form of:
> 
> ```java
> if (context instanceof TryLifecycleContext) {
> // ...
> }
> 
> if (context instanceof PropertyLifecycleContext) {
> // ...
> }

What common methods would TLC and PLC need for your purposes?

Reply:
Well... I'm not quite sure why `TryLifecycleContext` has a `testInstance` method but `PropertyLifecycleContext` has `testInstances` (perhaps because properties execute multiple tries?).

If a common ancestor exposes a `testInstances` method, I could remove the `instanceof` checks in two methods in the extension class.
In the same vein, if a common ancestor also exposes a `targetMethod`, then I wouldn't need to handle separate hooks for `PropertyLifecycleContext` and `TryLifecycleContext`.

Not sure if that makes sense from Jqwik's API perspective. Maybe it's not even worth doing that.
I also have suggestions for the hooks themselves, but I will need to review the code to write down the things I have noticed (will do so during the week).

P.D.: Please let me know if this PR looks good, that way I will run some static analysis and will mark it as ready for merge. Thanks :)

Reply:
I'm just about to be off for vacation for two weeks. I didn't find the time to review the current state. When I get back I'll have a look at the PR and the potential change to the LifeCycle hierarchy. Sorry for the delay.

The good news: As soon as your PR is ready to deploy, I can deploy jqwik 1.7.4 first, so that jqwik-micronaut could have its first non snapshot release directly afterwards.

Reply:
> Well... I'm not quite sure why `TryLifecycleContext` has a `testInstance` method but `PropertyLifecycleContext` has `testInstances` (perhaps because properties execute multiple tries?).

`PropertyLifecycleContext` also has `testInstance()`. The point of `testInstances()` is for nested test container classes. There you have several instances: from most inside class to top-level class.